# Penalized logistic regression

In this notebook, you will discover how to use penalized linear regression Lasso (L1), Ridge (L2) and Elasticnet (L1 + L2).

These penalties integrated to the cost function will help you train less complex models to avoid overfitting.

# Packages importation

In [ ]:
# Importation of the data for our classification example
from sklearn.datasets import load_breast_cancer

# Importation of the function to standardize the data
from sklearn.preprocessing import StandardScaler

# Importation of the train_test_split function which split randomly our data 
# into a train and test set
from sklearn.model_selection import train_test_split

# Importation of the logistic regression algorithm
from sklearn.linear_model import LogisticRegression

# Importation of the performance metrics
from sklearn.metrics import accuracy_score, precision_recall_curve, f1_score, roc_auc_score, roc_curve, confusion_matrix

# Importation of the maplotlib package to create graphics
import matplotlib.pyplot as plt

# Importation of numpy to use of vectors, matrices, tensors.
import numpy as np 

#Data Importation

In [ ]:
# Data frame for our classification
breast_cancer = load_breast_cancer()
X_classif = breast_cancer.data[:, ]
y_classif = breast_cancer.target

Use the Sklearn function *train_test_split* to split your dataset into two random set.

Use a random_state of 123 and use 10% of your dataset for the test set.

Feel free to use the [doc](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).

In [ ]:
# Use the function train_test_split to create your train and test set
X_train_classif, X_test_classif, y_train_classif, y_test_classif = train_test_split(X_classif, y_classif, 
                                                                    test_size=0.10, 
                                                                    random_state=123)

# Step 1 : Data standardization

For the use of a linear model it is essential to go through a step of normalization of the data.

This step allows to make the model interpretable but also to facilitate the convergence of the model.

Feel free to use the [doc](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html).

In [ ]:
# Initialize the StandardScaler function
scaler = StandardScaler()

# Fit the StandardScaler on the trainig set
scaler.fit(X_train_classif)

# Standardization of the training set
X_train_classif_norm = scaler.transform(X_train_classif)

# Standardization of the validation set
X_test_classif_norm = scaler.transform(X_test_classif)

In [ ]:
print('Mean of the training set : '+str(X_train_classif_norm.mean(axis=0)))
print('Standard deviation of the training set : '+str(X_train_classif_norm.std(axis=0)))

print('Mean of the testing set : '+str(X_test_classif_norm.mean(axis=0)))
print('Standard deviation of the testing set : '+str(X_test_classif_norm.std(axis=0)))

Mean of the training set : [-5.13911830e-16 -3.05745013e-16  8.53483950e-16  1.10003152e-15
  1.54650598e-15  7.12049777e-16 -3.42607887e-17  1.31622144e-16
 -4.95350289e-15  7.66292411e-15  1.53913340e-15  1.52915874e-15
  2.89265140e-16 -2.73435788e-16  1.46914815e-15  1.42854478e-15
  1.64798730e-17 -4.59701721e-16 -1.22119112e-15 -1.23165367e-16
  7.01695646e-16  3.83720833e-15 -9.50628465e-16  2.82326246e-16
  4.31078784e-15 -3.93619599e-16  2.29850861e-16  1.19966970e-16
 -4.12918397e-15  3.31635761e-15]
Standard deviation of the training set : [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1.]
Mean of the testing set : [ 0.05124296 -0.18874973  0.03751984  0.03937833 -0.08482864 -0.18283252
 -0.07809671 -0.05164212 -0.19793932 -0.29018854  0.16691644 -0.05227523
  0.14396937  0.1740289   0.00614669 -0.25922462 -0.16597387 -0.12024983
 -0.11072004 -0.29499964  0.04435633 -0.07442118  0.0294974   0.01545994
 -0.02918828 -0.1089334  -0.0558

Answers expected :

Mean of the training set : [-5.13911830e-16 -3.05745013e-16  8.53483950e-16  1.10003152e-15
  1.54650598e-15  7.12049777e-16 -3.42607887e-17  1.31622144e-16
 -4.95350289e-15  7.66292411e-15  1.53913340e-15  1.52915874e-15
  2.89265140e-16 -2.73435788e-16  1.46914815e-15  1.42854478e-15
  1.64798730e-17 -4.59701721e-16 -1.22119112e-15 -1.23165367e-16
  7.01695646e-16  3.83720833e-15 -9.50628465e-16  2.82326246e-16
  4.31078784e-15 -3.93619599e-16  2.29850861e-16  1.19966970e-16
 -4.12918397e-15  3.31635761e-15]


Standard deviation of the training set : [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1.]


Mean of the testing set : [ 0.05124296 -0.18874973  0.03751984  0.03937833 -0.08482864 -0.18283252
 -0.07809671 -0.05164212 -0.19793932 -0.29018854  0.16691644 -0.05227523
  0.14396937  0.1740289   0.00614669 -0.25922462 -0.16597387 -0.12024983
 -0.11072004 -0.29499964  0.04435633 -0.07442118  0.0294974   0.01545994
 -0.02918828 -0.1089334  -0.05586924 -0.0412101  -0.06118219 -0.18438358]

 
Standard deviation of the testing set : [0.91480598 0.66909629 0.90978263 1.03432425 0.99367801 0.87525274
 0.90633893 0.90440753 0.71909467 0.84083134 1.57807166 0.87352948
 1.60204167 1.78225409 1.01625717 0.75721494 0.65688383 0.93341027
 0.99874551 0.56807441 0.84475039 0.80330342 0.84141439 0.89898207
 0.95859295 1.1139185  1.010088   0.95613581 1.03214445 1.24999311]

# Step 2 : Model initialization

In the case of regression, there is no choice of hyperparameter.

It is therefore sufficient to just initialize the function.

Feel free to use the [doc](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html).

In [ ]:
reg = LogisticRegression()

In the case of lasso regression, you have to choose a value for alpha.

Alpha will control the regularization of the model.

$ J(w) =  \frac{1}{2m}[\frac{1}{C}\sum^m_{i=1}(\hat{y}^{(i)}-y^{(i)})^2+\sum^n_{j=1}|w_j|]$ 

For logistic regression in Sklearn you will use the same function for all penalization and just precise the type of penalty you want with the parameter *penalty*.

For this example initialize the regression with an alpha of 0.8, solver='saga' and a random_state of 123.

Feel free to use the [doc](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html).

In [ ]:
lasso = LogisticRegression(penalty='l1', C=0.8, random_state=123, solver='saga')

In the case of Ridge regression, you have to choose a value for alpha.

Alpha will control the regularization of the model.

$ J(w) =  \frac{1}{2m}[\frac{1}{C}\sum^m_{i=1}(\hat{y}^{(i)}-y^{(i)})^2+\sum^n_{j=1}w_j^2]$ 

For logistic regression in Sklearn you will use the same function for all penalization and just precise the type of penalty you want with the parameter *penalty*.

For this example initialize the regression with an alpha of 0.8 and a random_state of 123.

Feel free to use the [doc](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html).

In [ ]:
ridge = LogisticRegression(penalty='l2', C=0.8, random_state=123)

In the case of elasticnet regression, you have to choose a value for alpha and ratio.

Alpha will control the regularization of the model.

ratio is the mixing parameter beween lasso (ratio=0) and ridge (ratio=1)

$ J(w) =  \frac{1}{2m}[\frac{1}{C}\sum^m_{i=1}(\hat{y}^{(i)}-y^{(i)})^2+\frac{1-ratio}{2}\sum^n_{j=1}w_j^2 + ratio\sum^n_{j=1}|w_j|]$ 

For logistic regression in Sklearn you will use the same function for all penalization and just precise the type of penalty you want with the parameter *penalty*.

For this example initialize the regression with an alpha of 0.8, a ratio of 0.5, solver equal saga and a random_state of 123.

Feel free to use the [doc](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html).

In [ ]:
elasticnet = LogisticRegression(penalty='elasticnet', C=0.8, l1_ratio=0.5, 
                                random_state=123, solver='saga')

# Step 3 : Model training

You must train the four models.

In [ ]:
# Classic linear regression
reg.fit(X_train_classif_norm, y_train_classif)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [ ]:
# Lasso regression
lasso.fit(X_train_classif_norm, y_train_classif)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


LogisticRegression(C=0.8, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l1',
                   random_state=123, solver='saga', tol=0.0001, verbose=0,
                   warm_start=False)

In [ ]:
# Ridge regression
ridge.fit(X_train_classif_norm, y_train_classif)

LogisticRegression(C=0.8, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=123, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [ ]:
# ElasticNet regression
elasticnet.fit(X_train_classif_norm, y_train_classif)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


LogisticRegression(C=0.8, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=0.5, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='elasticnet',
                   random_state=123, solver='saga', tol=0.0001, verbose=0,
                   warm_start=False)

# Step 4 : Model validation

Your model is now trained, use it to predict the probability for your training and testing set for the four models.

In [ ]:
# Classic linear regression one hot prediction
x_train_reg_prediction = reg.predict(X_train_classif_norm)

x_test_reg_prediction = reg.predict(X_test_classif_norm)

# Classic linear regression probability 
x_train_reg_prediction_proba = reg.predict_proba(X_train_classif_norm)[:, 1]

x_test_reg_prediction_proba = reg.predict_proba(X_test_classif_norm)[:, 1]

In [ ]:
# Lasso regression
x_train_lasso_prediction = lasso.predict(X_train_classif_norm)

x_test_lasso_prediction = lasso.predict(X_test_classif_norm)

# Classic linear regression probability 
x_train_lasso_prediction_proba = lasso.predict_proba(X_train_classif_norm)[:, 1]

x_test_lasso_prediction_proba = lasso.predict_proba(X_test_classif_norm)[:, 1]

In [ ]:
# Ridge regression
x_train_ridge_prediction = ridge.predict(X_train_classif_norm)

x_test_ridge_prediction = ridge.predict(X_test_classif_norm)

# Classic linear regression probability 
x_train_ridge_prediction_proba = ridge.predict_proba(X_train_classif_norm)[:, 1]

x_test_ridge_prediction_proba = ridge.predict_proba(X_test_classif_norm)[:, 1]

In [ ]:
# ElasticNet regression
x_train_elasticnet_prediction = elasticnet.predict(X_train_classif_norm)

x_test_elasticnet_prediction = elasticnet.predict(X_test_classif_norm)

# Classic linear regression probability 
x_train_elasticnet_prediction_proba = elasticnet.predict_proba(X_train_classif_norm)[:, 1]

x_test_elasticnet_prediction_proba = elasticnet.predict_proba(X_test_classif_norm)[:, 1]

Compute the AUC for each model

In [ ]:
# Classic linear regression
auc_train = roc_auc_score(y_train_classif, x_train_reg_prediction_proba)

auc_test = roc_auc_score(y_test_classif, x_test_reg_prediction_proba)

print('AUC for the training set : '+str(auc_train))

print('AUC for the testing set : '+str(auc_test))

AUC for the training set : 0.9989821381665354
AUC for the testing set : 0.9797979797979799


In [ ]:
# Lasso regression
auc_train = roc_auc_score(y_train_classif, x_train_lasso_prediction_proba)

auc_test = roc_auc_score(y_test_classif, x_test_lasso_prediction_proba)

print('AUC for the training set : '+str(auc_train))

print('AUC for the testing set : '+str(auc_test))

AUC for the training set : 0.9984732072498029
AUC for the testing set : 0.9810606060606061


In [ ]:
# Ridge regression
auc_train = roc_auc_score(y_train_classif, x_train_ridge_prediction_proba)

auc_test = roc_auc_score(y_test_classif, x_test_ridge_prediction_proba)

print('AUC for the training set : '+str(auc_train))

print('AUC for the testing set : '+str(auc_test))

AUC for the training set : 0.9989493039138428
AUC for the testing set : 0.9797979797979799


In [ ]:
# ElasticNet regression
auc_train = roc_auc_score(y_train_classif, x_train_elasticnet_prediction_proba)

auc_test = roc_auc_score(y_test_classif, x_test_elasticnet_prediction_proba)

print('AUC for the training set : '+str(auc_train))

print('AUC for the testing set : '+str(auc_test))

AUC for the training set : 0.9985388757551878
AUC for the testing set : 0.9797979797979799


# Step 5 : Impact of the regularization term on the coefficient

Impact of the regularization term for the Lasso regression.

In [ ]:
for alpha_values in [0.01, 0.1, 0.2, 0.5, 1] :
  lasso = LogisticRegression(penalty='l1', C=alpha_values, 
                                random_state=123, solver='saga')
  lasso.fit(X_train_classif_norm, y_train_classif)
  print('Alpha = '+str(alpha_values))
  print(lasso.coef_)

Alpha = 0.01
[[ 0.          0.          0.          0.          0.          0.
   0.          0.          0.          0.          0.          0.
   0.          0.          0.          0.          0.          0.
   0.          0.         -0.26424952  0.         -0.16705349  0.
   0.          0.          0.         -0.53127883  0.          0.        ]]
Alpha = 0.1
[[-0.14507022  0.         -0.12041656  0.          0.          0.
   0.         -0.51510292  0.          0.         -0.24093316  0.
   0.          0.          0.          0.          0.          0.
   0.          0.         -0.91869054 -0.67675036 -0.70933674 -0.38701145
  -0.26333977  0.         -0.09230818 -0.83571792 -0.18593369  0.        ]]
Alpha = 0.2
[[-0.32047349 -0.1616604  -0.26753238 -0.2378532   0.          0.
   0.         -0.4705726   0.          0.         -0.51161577  0.
   0.         -0.17075759  0.          0.          0.          0.
   0.          0.02929312 -0.9082276  -0.79872879 -0.71663174 -0.61963469
  -

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the 

Impact of the regularization term for the Ridge regression.

In [ ]:
for alpha_values in [0.01, 0.1, 1, 10] :
  ridge = LogisticRegression(penalty='l2', C=alpha_values, 
                                random_state=123, solver='saga')
  ridge.fit(X_train_classif_norm, y_train_classif)
  print('Alpha = '+str(alpha_values))
  print(ridge.coef_)

Alpha = 0.01
[[-0.22332807 -0.18183998 -0.22109315 -0.21005999 -0.08444644 -0.09918108
  -0.16677098 -0.21736004 -0.06543112  0.08783398 -0.16256353  0.02234146
  -0.13821362 -0.15414772  0.01046901  0.02683516  0.02003664 -0.05175978
   0.04688977  0.0771214  -0.24694482 -0.22193547 -0.23774733 -0.21891494
  -0.1751445  -0.1403803  -0.18043599 -0.24268607 -0.16691238 -0.07579839]]
Alpha = 0.1
[[-0.40740949 -0.41973909 -0.39714213 -0.39459714 -0.12862023 -0.02642965
  -0.374073   -0.43650351 -0.07770733  0.24929369 -0.44283409  0.0979864
  -0.31796138 -0.38646379 -0.09230164  0.24820669  0.03532078 -0.07610843
   0.16401597  0.2510048  -0.51996549 -0.59656969 -0.47592709 -0.46861481
  -0.44791095 -0.15665607 -0.41588697 -0.4977698  -0.43455213 -0.14280643]]
Alpha = 1
[[-0.56050357 -0.61516366 -0.54299684 -0.58942542 -0.14051486  0.17581356
  -0.72922041 -0.73140609 -0.06449194  0.43366429 -0.91011559  0.30619026
  -0.60449697 -0.79908697 -0.29597873  0.66939875 -0.01959067 -0.10291682


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


Impact of the regularization term and the l1_ratio for the elasticnet regression.

In [ ]:
for alpha_values, ratio_values in zip([0.01, 0.01, 0.1, 0.5, 1, 1], [0, 1, 0.5, 0.5, 1, 0]) :
  elasticnet = LogisticRegression(penalty='elasticnet', C=alpha_values, l1_ratio=ratio_values, 
                                random_state=123, solver='saga')
  elasticnet.fit(X_train_classif_norm, y_train_classif)
  print('Alpha = '+str(alpha_values))
  print('ratio_values = '+str(ratio_values))
  print(elasticnet.coef_)

Alpha = 0.01
ratio_values = 0
[[-0.22332807 -0.18183998 -0.22109315 -0.21005999 -0.08444644 -0.09918108
  -0.16677098 -0.21736004 -0.06543112  0.08783398 -0.16256353  0.02234146
  -0.13821362 -0.15414772  0.01046901  0.02683516  0.02003664 -0.05175978
   0.04688977  0.0771214  -0.24694482 -0.22193547 -0.23774733 -0.21891494
  -0.1751445  -0.1403803  -0.18043599 -0.24268607 -0.16691238 -0.07579839]]
Alpha = 0.01
ratio_values = 1
[[ 0.          0.          0.          0.          0.          0.
   0.          0.          0.          0.          0.          0.
   0.          0.          0.          0.          0.          0.
   0.          0.         -0.26424952  0.         -0.16705349  0.
   0.          0.          0.         -0.53127883  0.          0.        ]]
Alpha = 0.1
ratio_values = 0.5
[[-0.35273939 -0.24863102 -0.33070369 -0.29041355  0.          0.
  -0.10552724 -0.42305855  0.          0.         -0.33737042  0.
  -0.0894243  -0.18854359  0.          0.          0.          0.

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the 